nest_elephant_tvb/translation/mpi_translator.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2020 Forschungszentrum Jülich GmbH and Aix-Marseille Université
+# "Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements; and to You under the Apache License, Version 2.0. "
+import time
+import numpy as np
+from mpi4py import MPI
+from threading import Thread, Lock
+from nest_elephant_tvb.translation.science_nest_to_tvb import store_data,analyse_data
+
+lock_status=Lock() # locker for manage the transfer of data from thread
+
+def init(path, param, comm, comm_receiver, comm_sender, loggers):
+    '''
+    Initialize the translation with MPI. NOTE: more information will be added with more changes. Currently the multithreaded version is still in use.
+    
+    '''
+    
+    # destructure logger list to indivual variables
+    logger_master, logger_receive, logger_send = loggers
+    # science part, see import
+    # TODO: use os.path (or similar) for proper file handling.
+    store = store_data(path+'/log/',param)
+    analyse = analyse_data(path+'/log/',param)
+    # variable for communication between thread
+    status_data=[0] # status of the buffer
+    
+    ############ NEW Code: MPI intracommunicator
+    # receive on rank 0, science and send on rank 1
+    # TODO: future work: mpi parallel, use rank 1-x for science and sending
+    # NOTE: still copy-paste from kim master thesis code....
+    # new server-intracommunicator without receiving rank 0
+    intracomm = comm.Create(comm.Get_group().Excl([0]))
+    # create the shared memory block / databuffer
+    buf_pkgs = 100 # NOTE: size hardcoded!
+    databuffer = _shared_mem_buffer(buf_pkgs,100,comm)
+    ############# NEW Code end
+    
+    # TODO:
+    # 1) receive package from nest, store in shared mem buffer
+    # 2) take from shared mem buffer, do science, send to tvb
+    # solve issues:
+    # - one package incoming, one package outgoing (synchronize)
+    # - mark buffer if new package arrived --> not to slow not to fast 
+    # - FIRST idea: 
+    #   - Receiving rank 0 check for 'NaN' in buffer, i.e. is it empty?
+    #   - if NaN then fill with new package
+    #   - Sending rank 1 (or more in the future) check for new package in buffer (not NaN)
+    #   - do science and send to TVB
+    #   - set buffer to NaN to mark as ready
+    # TODO: how to check for END of simulation?
+    
+    ############ OLD Code to replace: with threads, current solution
+    initialisation =np.load(param['init']) # initialisation value
+    buffer=[initialisation] # buffer of the rate to send it
+        
+    # create the thread for receive and send data
+    th_receive = Thread(target=receive, args=(logger_receive,store,status_data,buffer, comm_receiver))
+    th_send = Thread(target=send, args=(logger_send,analyse,status_data,buffer, comm_sender))
+
+    # start the threads
+    # FAT END POINT
```

Comment:
mix of using FAT END POINT and RICH END POINT at different places.

Reply:
see comment above!